hazelcast/src/main/java/com/hazelcast/instance/DefaultNodeExtension.java

```diff
@@ -219,13 +229,17 @@ public void shutdown() {
     }
 
     @Override
-    public void validateJoinRequest() {
+    public void validateJoinRequest(JoinMessage joinMessage) {
+        // check joining member's major.minor version is same as current cluster version's major.minor numbers
+        if (MAJOR_MINOR_VERSION_COMPARATOR.compare(joinMessage.getVersion(), node.getClusterService().getClusterVersion()) != 0) {
+            throw new VersionMismatchException("Joining node's version " + joinMessage.getVersion() + " is not compatible with "
+                    + node.getVersion());
+        }
```

Comment:
This constraint is something that we're introducing right now. Earlier, if by any chance two minor releases were compatible, the users might have tried to do a rolling-upgrade and it might have worked.
Do we want to have this constraint considering that the versioned serialization won't work in OS anyway?
CC: @jerrinot
(posting again so that it's not seen in outdated view)